haskell-lsp-types/src/Language/Haskell/LSP/Types/Uri.hs

```diff
@@ -113,3 +142,48 @@ platformAdjustToUriPath systemOS srcPath
     unescaped c
       | systemOS == windowsOS = isUnreserved c || c `elem` [':', '\\', '/']
       | otherwise = isUnreserved c || c == '/'
+
+-- | Newtype wrapper around FilePath that always has normalized slashes.
+-- The NormalizedUri and hash of the FilePath are cached to avoided
+-- repeated normalisation when we need to compute them (which is a lot).
+--
+-- This is one of the most performance critical parts of ghcide, do not
+-- modify it without profiling.
+data NormalizedFilePath = NormalizedFilePath NormalizedUri !FilePath
+    deriving (Generic, Eq, Ord)
+
+instance NFData NormalizedFilePath
+
+instance Binary NormalizedFilePath where
+  put (NormalizedFilePath _ fp) = put fp
+  get = do
+    v <- Data.Binary.get :: Get FilePath
+    return (toNormalizedFilePath v)
+
+instance Show NormalizedFilePath where
+  show (NormalizedFilePath _ fp) = "NormalizedFilePath " ++ show fp
+
+instance Hashable NormalizedFilePath where
+  hash (NormalizedFilePath uri _) = hash uri
+
+instance IsString NormalizedFilePath where
+    fromString = toNormalizedFilePath
+
+toNormalizedFilePath :: FilePath -> NormalizedFilePath
+toNormalizedFilePath fp = NormalizedFilePath nuri nfp
+  where nfp | fp == "" = "" 
+            -- ghcide want to keep empty paths instead of normalising them to "."
+            | otherwise = FP.normalise fp
+        uriStr = fileScheme <> "//" <> platformAdjustToUriPath System.Info.os nfp
+        nuriStr = T.pack $ escapeURIString isUnescapedInURI $ unEscapeString $ uriStr
```

Comment:
@cocreature i think this part is making roundtrip tests fail and i am not sure it is intentional. Consider the repl session:
```
> toNormalizedFilePath "C:\\a#s"
NormalizedFilePath "C:\\a#s"
> normalizedFilePathToUri $ toNormalizedFilePath "C:\\a#s"
NormalizedUri 1827512744 "file:///C:/a#s"
> uriToNormalizedFilePath $ normalizedFilePathToUri $ toNormalizedFilePath "C:\\a#s"
Just NormalizedFilePath "C:\\a"
```
shouldnt `NormalizedUri _"file:///C:/a#s"` be `NormalizedUri _"file:///C:/a%23s"`?
Like the old functions already does:
```
> filePathToUri "C:\a#s"
Uri {getUri = "file:///C:/%07%23s"}
> uriToFilePath $ filePathToUri "C:\\a#s"
Just "C:\\a#s"
```

Reply:
I've removed temporary the extra unescape/escape uri step to check if tests pass